scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/AsyncFlowDefRunner.scala

```diff
@@ -68,14 +73,51 @@ object AsyncFlowDefRunner {
 }
 
 /**
+ * This holds an internal thread to run
  * This holds an internal thread to submit run
  * a Config, Mode, FlowDef and return a Future holding the
  * JobStats
  */
 class AsyncFlowDefRunner extends Writer { self =>
   import AsyncFlowDefRunner._
 
-  private[this] val filesToCleanup = mutable.Set[String]()
+  private[this] val mutex = new AnyRef
+
+  private case class State(
+    filesToCleanup: Map[Mode, Set[String]],
+    forcedPipes: Map[(Config, Mode, TypedPipe[Any]), Future[TypedPipe[Any]]]) {
+
+    def addFilesToCleanup(m: Mode, s: Option[String]): State =
+      s match {
+        case Some(path) =>
+          val newFs = filesToCleanup.getOrElse(m, Set.empty[String]) + path
+          copy(filesToCleanup = filesToCleanup + (m -> newFs))
+        case None => this
+      }
+
+    def addPipe[T](c: Config,
+      m: Mode,
+      init: TypedPipe[T],
+      p: Future[TypedPipe[T]]): Option[State] =
+
+      forcedPipes.get((c, m, init)) match {
+        case None =>
+          Some(copy(forcedPipes = forcedPipes + ((c, m, init) -> p)))
+        case Some(exists) => None
+      }
+  }
+
+  private[this] var state: State = State(Map.empty, Map.empty)
+
+  private def updateState[S](fn: State => (State, S)): S =
+    mutex.synchronized {
+      val (st1, s) = fn(state)
+      state = st1
+      s
+    }
+  private def getState: State =
+    updateState { s => (s, s) }
```

Comment:
can you be concrete about how you want to change this? These is only one variable here, the state and you are copying that to be the result. Can you write the code you want me to replace it with because I don't follow your comment.